tree/dataframe/src/RDFSnapshotHelpers.cxx

```diff
@@ -930,3 +995,260 @@ ROOT::Internal::RDF::UntypedSnapshotRNTupleHelper::MakeNew(void *newName)
+   /// tree into the output file.
+   void RegisterBranch(std::string const &branchName, unsigned int variationIndex)
+   {
+      if (auto it = fBranchToVariationMapping.find(branchName); it != fBranchToVariationMapping.end()) {
+         if (variationIndex != it->second) {
+            throw std::logic_error("Branch " + branchName + " is being registered with different variation index.");
+         }
+         return;
+      }
+
+      // Neither branch nor systematic are known, so a new entry needs to be created
+      fNBits = std::max(fNBits, variationIndex);
+      const auto vectorIndex = variationIndex / 64u;
+      const auto bitIndex = variationIndex % 64u;
+
+      // Create bitmask branches as long as necessary to capture the bit
+      while (vectorIndex >= fBitMasks.size()) {
+         std::string bitmaskBranchName =
+            std::string{"R_rdf_mask_"} + fTree->GetName() + '_' + std::to_string(fBitMasks.size());
+         fBitMasks.push_back(Bitmask{bitmaskBranchName});
+         fTree->Branch(bitmaskBranchName.c_str(), fBitMasks.back().branchBuffer.get());
+      }
+
+      fBranchToVariationMapping[branchName] = variationIndex;
+      fBranchToBitmaskMapping[branchName] = std::make_pair(fBitMasks[vectorIndex].branchName, bitIndex);
+   }
+
+   /// Clear all bits, as if none of the variations passed its filter.
+   void ClearMaskBits()
+   {
+      for (auto &mask : fBitMasks)
+         mask.bitset.reset();
+   }
+
+   /// Set a bit signalling that the variation at `index` passed its filter.
+   void SetMaskBit(unsigned int index)
+   {
+      const auto vectorIndex = index / 64;
+      const auto bitIndex = index % 64;
+      fBitMasks[vectorIndex].bitset.set(bitIndex, true);
+   }
+
+   /// Test if any of the mask bits are set.
+   bool MaskEmpty() const
+   {
+      return std::none_of(fBitMasks.begin(), fBitMasks.end(), [](Bitmask const &mask) { return mask.bitset.any(); });
+   }
+
+   /// Write the current event and the bitmask to the output dataset.
+   void Write() const
+   {
+      if (!fTree)
+         throw std::runtime_error("The TTree associated to the Snapshot action doesn't exist, any more.");
+
+      for (auto const &mask : fBitMasks) {
+         *mask.branchBuffer = mask.bitset.to_ullong();
+      }
+
+      fTree->Fill();
+   }
+};
+
+ROOT::Internal::RDF::SnapshotHelperWithVariations::SnapshotHelperWithVariations(
+   std::string_view filename, std::string_view dirname, std::string_view treename, const ColumnNames_t &vbnames,
+   const ColumnNames_t &bnames, const RSnapshotOptions &options, std::vector<bool> &&isDefine,
+   ROOT::Detail::RDF::RLoopManager *outputLoopMgr, ROOT::Detail::RDF::RLoopManager *inputLoopMgr,
+   const std::vector<const std::type_info *> &colTypeIDs)
+   : fOptions(options), fInputLoopManager{inputLoopMgr}, fOutputLoopManager{outputLoopMgr}
+{
+   EnsureValidSnapshotTTreeOutput(fOptions, std::string(treename), std::string(filename));
+
+   TFile::TContext fileCtxt;
```

Comment:
I usually see `TDirectory::TContext` being used in such cases, any reason to use `TFile::TContext` here?